### PR TITLE
Fix dangling ID when `pub use`ing item which is Doc(hidden) or inherits it in rustdoc JSON output

### LINKED ISSUE #117718: rustdoc-json: Danging ID when `pub use`ing item from `#[doc(hidden)]` module
With the following code:
```rust
#[doc(hidden)]
pub mod __secrets {
    pub struct Foo;
}

pub use __secrets::Foo;
```
It has a dangling ID:
```
---- [rustdoc-json] tests/rustdoc-json/reexport/pub_use_from_doc_hidden_mod.rs stdout ----

error: jsondoclint failed!
status: exit status: 1
command: "/home/gh-aDotInTheVoid/rust2/build/aarch64-unknown-linux-gnu/stage0-tools-bin/jsondoclint" "/home/gh-aDotInTheVoid/rust2/build/aarch64-unknown-linux-gnu/test/rustdoc-json/reexport/pub_use_from_doc_hidden_mod/pub_use_from_doc_hidden_mod.json"
stdout: none
--- stderr -------------------------------
0:4:1750 not in index or paths, but referred to at '$.index["0:6-0:4:1750"].inner.import.id'
Error: Errors validating json /home/gh-aDotInTheVoid/rust2/build/aarch64-unknown-linux-gnu/test/rustdoc-json/reexport/pub_use_from_doc_hidden_mod/pub_use_from_doc_hidden_mod.json
------------------------------------------
```

As it has the following json (redacted)

```json
  "index": {
    "0:0:1751": {
      "inner": {"module": {"is_crate": true, "is_stripped": false, "items": ["0:6-0:4:1750"]}},
      "name": "pub_use_from_doc_hidden_mod"
    },
    "0:6-0:4:1750": {
      "inner": {"import": {"glob": false, "id": "0:4:1750", "name": "Foo", "source": "__secrets::Foo"}},
      "name": null
    },
```

Version: 28acba3c61246ba5974df332a064ab0c19b1cff8

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Fixes https://github.com/rust-lang/rust/issues/117718.

I changed a bit where we actually filter out some reexports to be done directly in the JSON conversion. When the reexport is public but the reexported item is doc hidden, it generates the reexport but with a `null` ID.

r? @aDotInTheVoid

### CODE REVIEW

**Comment by GuillaumeGomez** [comment]
I rebased and updated the failing test (it was testing something changed in this PR).

**Comment by obi1kenobi** [comment]
I have a serious concern here, and a request to reconsider the interpretation of how `doc(hidden)` applies across module boundaries.

Let's look at this example from the doc comment:
```rust
#[doc(hidden)]
pub mod foo {
    pub struct Bar;
}

pub use foo::Bar;
```

To me, this seems like a public re-export of an intended-to-be-public item. In other words, when I see this code, I expect its `doc(hidden)` semantics to be that it is the _module_ that is hidden, but non-hidden ways to access its contents are public API and "fair game". In my interpretation, if the programmer intended that to not be the case, they could have hidden the re-export (or `pub struct Bar` itself).

If `pub struct Bar` is always considered hidden regardless of re-exports (whether themselves hidden or not), I think that has strange implications for what is and is not public API. In that interpretation, the re-export is public API but the item it points to is not public API ... which is just confusing. It seems like it says "I make it part of the public API that this non-public-API type reserves a name in the public API" and nothing else. That's just strange! What public API operations are allowed over that re-export item? It feels to me like the answer is that the item is "neither here nor there" with respect to being public API or not.

It also seems to risk opening a terrible can of worms with further edge cases:
```rust
mod defn {
    pub struct Example;
}

#[doc(hidden)]
pub mod hide {
    pub use crate::Example as Example;
}

// Here, `Example` is public API.
pub use defn::Example;

// Replace it with the below code and `Example` is no longer public API.
// Worse, we can't distinguish between these two cases via rustdoc JSON,
// so this interpretation would permanently break cargo-semver-checks.
// pub use hide::Example;
```

For these reasons, I think a better interpretation of whether an item is `doc(hidden)` is one based on reachability: if you can reach the item without going through any `doc(hidden)` items (the item itself + any modules and re-exports), it is not hidden. If all paths require going through at least one hidden item, the item is hidden.

**Comment by GuillaumeGomez** [comment]
I followed the same existing logic as for public item in private module:

```rust
mod x {
    pub struct Y {}
}

/// a
pub use x::Y as Z;
```

So I agree with you, but then this would be a change to be done in another PR as it would change how we handle things completely.

**Comment by obi1kenobi** [comment]
Thanks! Just to double-check since it's important for cargo-semver-checks:

> this would be a change to be done in another PR as it would change how we handle things completely.

Am I correct in understanding that you feel the "reachability" interpretation I proposed should be how `doc(hidden)` works, and the team plans to adjust rustdoc in line with that in future PRs?

I'm asking because I wanted to release a new cargo-semver-checks version next week with the ability to ignore `doc(hidden)` items and no longer report false-positives when such items change in major breaking ways. However, its implementation of `doc(hidden)` handling is reachability-based since (as I mentioned in my earlier post), it's the only one that could work within the current rustdoc JSON system and is able to reuse the existing visibility logic. So I wanted to double-check whether those semantics are "deemed correct" so as not to falsely promise users that the new cargo-semver-checks would correctly handle `doc(hidden)` items. Right now, cargo-semver-checks outputs both private and hidden items so it's not affected by this PR specifically — which is why I'm more worried about the intended semantics than about when those changes are implemented.

**Comment by GuillaumeGomez** [comment]
I plan to at least open a discussion about how reexported items from private/hidden modules should appear in the JSON output. I'm not too happy with the current situation as it differs between the HTML and JSON outputs. The HTML output inline the items, showing all their information, whereas the JSON output just shows a reexport and we can't get the information of the reexported item.

So I suggest this: please open an issue on this repository and start a discussion on zulip so we can all agree exactly on what we want and once we reach a consensus (between rustdoc team members and cargo-semver-checks), we'll implement this change. Does it sound good to you?

And again: this PR is really about fixing a current bug while not changing the current situation. So it'd be nice to have it merged ahead of time so we can at least have tests to update when we change JSON output.

**Comment by obi1kenobi** [comment]
Absolutely. I'll open a separate issue and Zulip discussion :+1:

Thanks again!

File: src/librustdoc/json/conversions.rs [hunk 1]
```diff
@@ -760,20 +760,24 @@ impl FromWithTcx<clean::Discriminant> for Discriminant {
...
 impl FromWithTcx<clean::Import> for Import {
     fn from_tcx(import: clean::Import, tcx: TyCtxt<'_>) -> Self {
-        use clean::ImportKind::*;
-        let (name, glob) = match import.kind {
-            Simple(s) => (s.to_string(), false),
-            Glob => (
+        let (name, glob, id) = match import.kind {
+            ImportKind::Simple(s) => {
+                let is_from_doc_hidden =
+                    import.imported_item_is_doc_hidden(tcx) || import.inherits_doc_hidden(tcx);
```
Comment (aDotInTheVoid) [inline src/librustdoc/json/conversions.rs:766]: NIT: The definition of `inherits_doc_hidden` already calls `imported_item_is_doc_hidden`. That should be de-duplicated, either here or there.

File: tests/rustdoc-json/reexport/pub_use_doc_hidden.rs [hunk 2]
```diff
@@ -6,8 +6,9 @@ mod repeat_n {
     pub struct RepeatN {}
 }
 
-/// not here
+/// is here
```
Comment (aDotInTheVoid) [inline tests/rustdoc-json/reexport/pub_use_doc_hidden.rs:9]: I don't think this is the right behavior. If the underlying item isn't part of the public API than I don't think the import of it should be either.

File: tests/rustdoc-json/doc_hidden_reexports.rs [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+#[doc(hidden)]
+mod x {
+    pub struct Y {}
+}
+
+// @has "$.index[*].inner[?(@.import.name=='UsedHidden')].import.source" '"submodule::Hidden"'
+// @has "$.index[*].inner[?(@.import.name=='UsedHidden')].import.id" "null"
+pub use submodule::Hidden as UsedHidden;
+// @has "$.index[*].inner[?(@.import.name=='Z')].import.source" '"x::Y"'
+// @has "$.index[*].inner[?(@.import.name=='Z')].import.id" 'null'
```
Comment (aDotInTheVoid) [inline tests/rustdoc-json/doc_hidden_reexports.rs:18]: As discussed in <https://rust-lang.zulipchat.com/#narrow/stream/266220-t-rustdoc/topic/pub.20re-export.20of.20item.20from.20.60doc.28hidden.29.60.20module/near/401627083>, this import should be a part of the output, and so the ID should be present

**Review by aDotInTheVoid — CHANGES_REQUESTED** [review]

File: tests/rustdoc-json/reexport/pub_use_doc_hidden.rs (hunk 2 above)
Comment (GuillaumeGomez) [reply tests/rustdoc-json/reexport/pub_use_doc_hidden.rs:9]: The import is publicly accessible so I don't see why not?

File: tests/rustdoc-json/doc_hidden_reexports.rs (hunk 3 above)
Comment (aDotInTheVoid) [reply tests/rustdoc-json/doc_hidden_reexports.rs:18]: > I followed the same logic as for reexport of item inside private module.

When reexporting an item defined in a private module, the import will have the ID of the underlying item. We should do the same thing here.

> We discussed that after this PR, we want to inline such items

Please no. Inlining was a massive source of bugs for rustdoc-json, and removing it was a great improvement. I'm strongly opposed to bringing it back.

File: tests/rustdoc-json/reexport/pub_use_doc_hidden.rs (hunk 2 above)
Comment (aDotInTheVoid) [reply tests/rustdoc-json/reexport/pub_use_doc_hidden.rs:9]: 1. Ideologically: `RepeatN` has asked to not be part of the public API. We don't show it in the HTML docs.
2. Practicly: Theirs very little use to consumers telling them "You import something here, but you don't get to know what it it". If it's a part of the API, we should tell them the full story. If it isn't then we shouldn't acknowledge the import at all. What purpose does this weird in-between serve? (If theirs a compelling use case for just knowing this, that'd easily change my mind)

File: tests/rustdoc-json/doc_hidden_reexports.rs (hunk 3 above)
Comment (GuillaumeGomez) [reply tests/rustdoc-json/doc_hidden_reexports.rs:18]: Having an ID pointing to an item that is not available means that the current bug fixed by this PR will remain.

**Comment by GuillaumeGomez** [comment]
I think this PR is not the right way to solve this issue so closing it.

**Pull request closed without merging**